### ISSUE TITLE
module 'comfy.sample' has no attribute 'prepare_mask'

### ISSUE THREAD

**Issue by cardenluo** [comfyanonymous/ComfyUI#3247]
Error occurred when executing KSampler:

module 'comfy.sample' has no attribute 'prepare_mask'

File "F:\ComfyUI_Windows_portable\ComfyUI\execution.py", line 151, in recursive_execute
output_data, output_ui = get_output_data(obj, input_data_all)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "F:\ComfyUI_Windows_portable\ComfyUI\execution.py", line 81, in get_output_data
return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "F:\ComfyUI_Windows_portable\ComfyUI\custom_nodes\ComfyUI-0246\utils.py", line 381, in new_func
res_value = old_func(*final_args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "F:\ComfyUI_Windows_portable\ComfyUI\execution.py", line 74, in map_node_over_list
results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "F:\ComfyUI_Windows_portable\ComfyUI\nodes.py", line 1344, in sample
return common_ksampler(model, seed, steps, cfg, sampler_name, scheduler, positive, negative, latent_image, denoise=denoise)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "F:\ComfyUI_Windows_portable\ComfyUI\nodes.py", line 1314, in common_ksampler
samples = comfy.sample.sample(model, noise, steps, cfg, sampler_name, scheduler, positive, negative, latent_image,
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "F:\ComfyUI_Windows_portable\ComfyUI\custom_nodes\cardenc_input\modules\impact\sample_error_enhancer.py", line 9, in informative_sample
return original_sample(*args, **kwargs) # This code helps interpret error messages that occur within exceptions but does not have any impact on other operations.
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "F:\ComfyUI_Windows_portable\ComfyUI\custom_nodes\ComfyUI-AnimateDiff-Evolved\animatediff\sampling.py", line 292, in motion_sample
function_injections.inject_functions(model, params)
File "F:\ComfyUI_Windows_portable\ComfyUI\custom_nodes\ComfyUI-AnimateDiff-Evolved\animatediff\sampling.py", line 217, in inject_functions
self.orig_prepare_mask = comfy.sample.prepare_mask

**Comment by cardenluo** [comment]
![image](https://github.com/comfyanonymous/ComfyUI/assets/133877069/700b8f8b-8ab4-481e-9838-21009bb7352c)
copy from  all def…… into sample ,can run ok

**Comment by Zakutu** [comment]
Could you please send me the improved sample.py code just to be sure?

**Comment by Zakutu** [comment]
In my case, the above code worked fine.

> import torch import comfy.model_management import comfy.samplers import comfy.utils import numpy as np import ロギング
> 
> def prepare_noise(latent_image,シード,noise_inds=None): """ は、 潜在イメージとシードを指定してランダム ノイズを作成します。 オプションの引数スキップを使用すると、指定されたシードに対する x 回のノイズ生成をスキップして破棄できます。 """ ジェネレーター = torch.manual_seed(seed) noise_inds が None の場合: return torch.randn(latent_image.size(), dtype=latent_image.dtype,layout=latent_image.layout,generator=generator, device="cpu")
> 
> ```
> unique_inds, inverse = np.unique(noise_inds, return_inverse=True)
> noises = []
> for i in range(unique_inds[-1]+1):
>     noise = torch.randn([1] + list(latent_image.size())[1:], dtype=latent_image.dtype, layout=latent_image.layout, generator=generator, device="cpu")
>     if i in unique_inds:
>         noises.append(noise)
> noises = [noises[i] for i in inverse]
> noises = torch.cat(noises, axis=0)
> return noises
> ```
> 
> def prepare_sampling(model,noise_shape,positive,negative,noise_mask): logging.warning("警告:comfy.sample.prepare_sampling はもう使用されていないため、削除できます") return model,positive,negative,noise_mask,[]
> 
> def cleanup_Additional_models(models): logging.warning("警告: comfy.sample.cleanup_Additional_models はもう使用されていないため、削除できます")
> 
> def サンプル(モデル、ノイズ、ステップ、cfg、sampler_name、スケジューラ、ポジティブ、ネガティブ、潜在イメージ、denoise=1.0、disable_noise=False、start_step=None、last_step=None、force_full_denoise=False、noise_mask=None、sigmas=None、コールバック=None、disable_pbar=False、seed=None): sampler = comfy.samplers.KSampler(model、steps=steps、device=model.load_device、sampler=sampler_name、scheduler=scheduler、denoise=denoise、model_options=model.model_options)
> 
> ```
> samples = sampler.sample(noise, positive, negative, cfg=cfg, latent_image=latent_image, start_step=start_step, last_step=last_step, force_full_denoise=force_full_denoise, denoise_mask=noise_mask, sigmas=sigmas, callback=callback, disable_pbar=disable_pbar, seed=seed)
> samples = samples.to(comfy.model_management.intermediate_device())
> return samples
> ```
> 
> def sample_custom(モデル、ノイズ、cfg、サンプラー、シグマ、ポジティブ、ネガティブ、潜在イメージ、noise_mask=None、callback=None、disable_pbar=False、seed=None): サンプル = comfy.samplers.sample(モデル、ノイズ、ポジティブ、 negative、cfg、model.load_device、サンプラー、シグマ、model_options=model.model_options、latent_image=latent_image、denoise_mask=noise_mask、callback=コールバック、disable_pbar=disable_pbar、seed=seed) サンプル = tables.to(comfy.model_management.intermediate_device( )) サンプルを返す
> 
> def common_ksampler(モデル、シード、ステップ、cfg、sampler_name、スケジューラ、ポジティブ、ネガティブ、潜在イメージ、denoise=1.0、disable_noise=False、start_step=None、last_step=None、force_full_denoise=False、noise_mask=None、sigmas=None、コールバック=None、disable_pbar=False、custom=None): ノイズ = prepare_noise(latent_image、seed) return サンプル(モデル、ノイズ、ステップ、cfg、sampler_name、スケジューラ、ポジティブ、ネガティブ、latent_image、denoise、disable_noise、start_step、last_step、force_full_denoise) 、ノイズマスク、シグマ、コールバック、disable_pbar、シード)
> 
> def process_sample_state(pipe、samp_model、samp_clip、samp_samples、samp_vae、samp_seed、samp_positive、samp_negative、steps、start_step、last_step、cfg、sampler_name、scheduler、denoise、image_output、link_id、save_prefix、tile_size、prompt、extra_pnginfo、preview、 _latent、force_full_denoise 、disable_noise、samp_custom): return common_ksampler(samp_model、samp_seed、steps、cfg、sampler_name、scheduler、samp_positive、samp_negative、samp_samples、denoise、disable_noise、start_step、last_step、force_full_denoise、samp_custom)
> 
> def prepare_mask(): logging.warning("警告:comfy.sample.prepare_mask はもう使用されていないため、削除できます")
> 
> def prepare_Additional_models(models): logging.warning("警告:comfy.sample.prepare_Additional_models はもう使用されていないため、削除できます")
> 
> def get_Additional_models(): logging.warning("警告: comfy.sample.get_Additional_models はもう使用されていないため、削除できます")

**Comment by Zakutu** [comment]
In my case, the above code worked fine.

**Comment by cardenluo** [comment]
@Zakutu the error is : module 'comfy.sample' has no attribute 'prepare_mask' ;
so ,I find there is  def prepare_mask() in the sampler_helpers.py,   I just copy the code into comfy.sample.py.

However, I update latest comfyui ,there's no need to copy again.

**Comment by sauceruney** [comment]
I did all this above and I'm still getting the error. I'm running a new install of ComfyUI in StabilityMatrix. This happens whether I use Ksampler or the Advanced version.

Error occurred when executing KSamplerAdvanced:

module 'comfy.sample' has no attribute 'prepare_mask'

File "C:\Users\sauce\AppData\Roaming\StabilityMatrix\Packages\ComfyUI\execution.py", line 151, in recursive_execute
output_data, output_ui = get_output_data(obj, input_data_all)
File "C:\Users\sauce\AppData\Roaming\StabilityMatrix\Packages\ComfyUI\execution.py", line 81, in get_output_data
return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
File "C:\Users\sauce\AppData\Roaming\StabilityMatrix\Packages\ComfyUI\execution.py", line 74, in map_node_over_list
results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
File "C:\Users\sauce\AppData\Roaming\StabilityMatrix\Packages\ComfyUI\nodes.py", line 1378, in sample
return common_ksampler(model, noise_seed, steps, cfg, sampler_name, scheduler, positive, negative, latent_image, denoise=denoise, disable_noise=disable_noise, start_step=start_at_step, last_step=end_at_step, force_full_denoise=force_full_denoise)
File "C:\Users\sauce\AppData\Roaming\StabilityMatrix\Packages\ComfyUI\nodes.py", line 1314, in common_ksampler
samples = comfy.sample.sample(model, noise, steps, cfg, sampler_name, scheduler, positive, negative, latent_image,
File "C:\Users\sauce\AppData\Roaming\StabilityMatrix\Packages\ComfyUI\custom_nodes\ComfyUI-AnimateLCM\animatediff\sampling.py", line 264, in motion_sample
function_injections.inject_functions(model, params)
File "C:\Users\sauce\AppData\Roaming\StabilityMatrix\Packages\ComfyUI\custom_nodes\ComfyUI-AnimateLCM\animatediff\sampling.py", line 198, in inject_functions
self.orig_prepare_mask = comfy.sample.prepare_mask

I deleted and reinstalled AnimateDiffEvo and I'm still getting a ton of lora key not loaded warnings before all this happens. And a "UserWarning: 1Torch was not compiled with flash attention. (Triggered internally at ..\aten\src\ATen\native\transformers\cuda\sdp_utils.cpp:455.)"

**Comment by ltdrdata** [comment]
> prepare_mask

ComfyUI-AnimateLCM is so much outdated. And it isn't patched yet.
